Automated cherry pick of #129661: DRA CEL: add missing size estimator

pohly opened the pull request:
Cherry pick of #129661 on release-1.32.
Cherry pick of #129749 on release-1.32.

#129661: DRA CEL: add missing size estimator
#129749: DRA: CEL cost estimate during validation

For details on the cherry pick process, see the [cherry pick requests](https://git.k8s.io/community/contributors/devel/sig-release/cherry-picks.md) page.

```release-note
DRA: CEL expressions using attribute strings exceeded the cost limit because their cost estimation was incomplete. Cost estimation was unnecessarily also computed in the scheduler.
```

pohly committed:
DRA CEL: add missing size estimator

Not implementing a size estimator had the effect that strings retrieved from
the attributes were treated as "unknown size", leading to wildly overestimating
the cost and validation errors even for even simple expressions like this:

    device.attributes["qat.intel.com"].services.matches("[^a]?sym")

Maximum number of elements in maps and the maximum length of the driver name
string were also ignored resp. missing. Pre-defined types like
apiservercel.StringType must be avoided because they are defined as having
a zero maximum size.

pohly commented:
This is not an obvious candidate for cherry-picking (no security issue, no regression because the specific usage has never worked, and not critical because no data loss etc).

But it affects a beta feature that people wanted to use with 1.32 and the problem was something that got reported by different driver authors (Intel for QAT, NVIDIA for GPUs) because they ran into it when preparing driver releases for 1.32. So let's discuss whether it's worth to cherry-pick...

- Issue: https://github.com/kubernetes/kubernetes/pull/129661
- Scope of the change: limited to DRA CEL expressions, no impact on other CEL usage
- Risks: low, because once it was understood what needed to be done, it was possible to adapt the [corresponding code](https://github.com/kubernetes/kubernetes/blob/ab54e442c6cfc64d25462906c276950796e6803c/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/compilation.go#L313-L345) from admission control - @jpbetz, agreed?
- Tests: no changes in existing unit test cases, new cases confirm that the change works as intended
- Key stakeholder SIG reviewers/approvers: @klueska, @SergeyKanzhelev, please confirm for SIG Node that this can and should be backported

pohly commented:
/cc @liggitt 

Because he was concerned about the intent to backport and for API approval.

klueska commented:
One argument for backporting is that without this fix, the following (seemingly simple) CEL expression fails to evaluate:
```
 - cel:
            expression: |
              device.attributes['gpu.nvidia.com'].productName.lowerAscii().matches('^.*a100.*$')
```
This is a common way to select a specific type of GPU by its product name.

I can't speak to the complexity of the fix itself (and therefore the risk of backporting), but from a usability point of view, not backporting this means that users of DRA are artifically limited in the types of selection they can do on devices.

/cc @jpbetz to comment on the risk of backporting

pohly commented:
Not backporting also has the negative effect that those users who could give us feedback on writing those more complex CEL expressions would have to wait for 1.33. As seen with the promotion to beta in 1.32, it's not enough to have something available in the master branch. It needs to be released or almost released (= included in a tagged alpha release) before people start taking notice. For users there's the additional complication that they often depend on cloud or Kubernetes providers.

pohly commented:
It was pointed out that this used to work in early 1.32 alpha releases (@klueska had tested at that time) and then broke later when adding the CEL cost limit, so this *is* in fact a regression, at least when also considering alphas. This further strengthens the arguments that this should get cherry-picked.

liggitt commented:
> It was pointed out that this used to work in early 1.32 alpha releases (@klueska had tested at that time) and then broke later when adding the CEL cost limit

I don't really buy this... *any* cost limit added to the beta feature could be framed as a regression since it would make arbitrarily expensive expressions allowed in the alpha fail validation. We *intend* to reject *some* expressions that worked in alpha, the only question is which / how many.

> +1 to backporting this. This effectively makes it impossible to use some valid CEL expressions in prior minors. This was not our intent, and it "feels buggy" to have expressions that work on 1.33+ be rejected due to a mistake on expression costing.
> 
> I consider this low risk. The way the code change is made, only DRA is impacted, and it's easy to reason about the code and the test coverage serves as a strong safety net.

I think we're trying to weigh these risks against each other:
* Risk of alpha users being unable to use reasonable expressions due to overly aggressive cost enforcement. Clarifying how unusable the overly aggressive cost enforcement makes the feature would be helpful. Does "impossible to use some valid CEL expressions" mean *most* or *more complex*?
* Risk of regressing the beta feature in 1.32 with a non-trivial cost estimator change. Clarifying the risk here would be helpful. Worst-case, if the new cost estimation code errored / panicked, what would it break? Just the individual DRA API write request that would have failed due to cost estimation rejection anyway, or could that take down a server / controller?

johnbelamaric commented:
> I don't really buy this... _any_ cost limit added to the beta feature could be framed as a regression since it would make arbitrarily expensive expressions allowed in the alpha fail validation. We _intend_ to reject _some_ expressions that worked in alpha, the only question is which / how many.
> 

This issue though here is that it is *not* an arbitrarily expensive operation - the length is enforced and was enforced but is just not counted properly in the cost estimation. I get your point, but I do think that's a material difference from "any expression".

> * Risk of alpha users being unable to use reasonable expressions due to overly aggressive cost enforcement. Clarifying how unusable the overly aggressive cost enforcement makes the feature would be helpful. Does "impossible to use some valid CEL expressions" mean _most_ or _more complex_?

These are *beta* users not *alpha* users.

IIUC, any expression that either makes a comparison between two string attributes, or makes a function call on a string attribute (like a regex match) will fail validation here. In particular, that means no regex matching against these strings. I think the test cases only did comparisons of these strings to fixed length strings, so the those fixed length strings were used in the test case cost estimations.

I am not sure how smart the cost estimator is. It might be possible to "hack" around it by adding an additional, always true clause that limits the attribute string length. For example instead of `attribute.matches('.*')` you would put `attribute.length() < 100 && attribute.matches('.*')`. A human could see that here you can assume attribute is less than 100 characters and so estimate the `matches` cost. I don't know if the algorithm is that smart. It's also junk that will likely creep into CEL expressions long long into the future...

liggitt commented:
> These are beta users not alpha users.

Sorry... I meant users moving from alpha → beta losing ability to use expressions they did previously. New beta users new to the feature won't have any existing expressions they would consider the current state to regress.

I'd look to folks who know the DRA domain (@pohly, @johnbelamaric, @klueska) to describe how limiting this issue is for alpha → beta users, and the folks who know the cost estimator impl / wiring (@jpbetz, @pohly) to describe the risk to the whole cluster if the new code fails / panics in unexpected ways.

I mostly want to be super clear in the backport justification why we consider this 1) severe enough to backport and 2) low-risk enough to backport.

pohly commented:
My take (see also https://github.com/kubernetes/kubernetes/pull/129690#issuecomment-2598404158):

> I'd look to folks who know the DRA domain (@pohly, @johnbelamaric, @klueska) to describe how limiting this issue is for alpha → beta users

It prevents some advanced use cases. Two DRA driver teams ran into this independently while developing DRA drivers for 1.32 when trying to satisfy expected use cases for their drivers; those use cases won't be possible with 1.32 unless we backport.

> and the folks who know the cost estimator impl / wiring (@jpbetz, @pohly)

See https://github.com/kubernetes/kubernetes/pull/129690#issuecomment-2605375037 for @jpbetz's +1 for backporting. The risk is low because the only really "new" code isn't really that new - it was used before for admission control and only [slightly adapted](https://github.com/kubernetes/kubernetes/pull/129661#discussion_r1919641270) for DRA.

liggitt commented:
> The risk is low because the only "new" code isn't really that new - it was used before for admission control and only [slightly adapted](https://github.com/kubernetes/kubernetes/pull/129661#discussion_r1919641270) for DRA.

Can we bound the worst case risk by determining what the blast radius is if the new code errors or panics?

pohly commented:
> Can we bound the worst case risk by determining what the blast radius is if the new code errors or panics?

This runs during the scheduler's filter operation during scheduling of a pod which uses DRA. If it errors, scheduling that pod fails. If it panics, the scheduler would crash.

It also runs during validation in the apiserver.

I understand that this is a large blast radius. It's just not very likely.

Would it be more acceptable if the cherry-pick had additional code which catches panics and turns them into validation and scheduling errors?

liggitt commented:
> This runs during the scheduler's filter operation during scheduling of a pod which uses DRA. If it errors, scheduling that pod fails. If it panics, the scheduler would crash.

The cost estimation code is active in the scheduler? That doesn't seem correct (even in master). Shouldn't that *only* be active in the API validation side, where if it panicked it would prevent write of the DRA object containing the panic-triggering expression?

jpbetz commented on staging/src/k8s.io/dynamic-resource-allocation/cel/compile.go:
```diff
@@ -322,3 +352,67 @@ func (m mapper) Find(key ref.Val) (ref.Val, bool) {
+// Without this, the maximum string size of e.g. `device.attributes["dra.example.com"].services` would be unknown.
+//
+// sizeEstimator is derived from the sizeEstimator in k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel.
+type sizeEstimator struct {
+	compiler *compiler
+}
+
+func (s *sizeEstimator) EstimateSize(element checker.AstNode) *checker.SizeEstimate {
```
@liggitt for context, this is an exact copy of https://github.com/kubernetes/kubernetes/blob/3d2ee2fbb729857e5f71868107882f9e3c2aa43e/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/compilation.go#L313.  We have plans to follow up this PR and centralize the logic in a exportable way after this PR.  I mention it here to highlight that this is a well tested routine that we use in CRD validation rules.

liggitt replied:
> @liggitt for context, this is an exact copy of https://github.com/kubernetes/kubernetes/blob/3d2ee2fbb729857e5f71868107882f9e3c2aa43e/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/compilation.go#L313

is it? these bits look unique to this impl, and I don't have context to know how that impacts the logic

```
	switch path[0] {
	case deviceVar:
		currentNode = s.compiler.deviceType
	default:
		// Unknown root, shouldn't happen.
		return nil
	}
```

```
			if !ok {
				// If this is an attribute map, then we know that all elements
				// have the same maximum size as set in attributeType, regardless
				// of their name.
				if currentNode.ElemType == attributeType {
					currentNode = attributeType
					continue
				}
				return nil
			}
```

pohly replied:
That's the "adaption to DRA" that I mentioned earlier.

pohly commented:
> The cost estimation code is active in the scheduler?

`CompileCELExpression` sets a cost estimator unconditionally, so the new code is also active when used to prepare for evaluation in the scheduler. I copied that approach from here:
https://github.com/kubernetes/kubernetes/blob/25be4a4bd409249ce9f1289b3cd16b3119df80f4/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/compilation.go#L235-L249

Perhaps it doesn't matter there because checking the CEL expression and using it both happen in the apiserver and are done only once?

Skipping the cost estimation in the scheduler is indeed better. I'll fix on master and then can include that here.

>  Shouldn't that only be active in the API validation side, where if it panicked it would only prevent write of the DRA object containing the panic-triggering expression?

Wouldn't that then still take down the apiserver? `HandleCrash` logs, then continues to panic. Is the apiserver using some additional protections to continue running after a panic?

pohly committed:
DRA CEL: skip estimating the cost in the scheduler

Compiling a CEL expression used to do the cost estimation, whether the caller
needed the result or not. Now callers can skip it. The scheduler does that,
through the CEL cache.

The main advantage is that failures in the estimator (like panics) are limited
to the apiserver. Performance in the scheduler is not expected to benefit much
because compilation results are cached.

liggitt commented:
> > Shouldn't that only be active in the API validation side, where if it panicked it would only prevent write of the DRA object containing the panic-triggering expression?
> 
> Wouldn't that then still take down the apiserver? `HandleCrash` logs, then continues to panic. Is the apiserver using some additional protections to continue running after a panic?

No, panics within the main goroutine of an http handler get recovered and do not exit the server

pohly commented:
This PR now also includes https://github.com/kubernetes/kubernetes/pull/129749 and thus limits the blast radius to the apiserver validation, which is protected against panics as @liggitt pointed out.

pohly commented:
I also tested that only the writes are affected. I first created a DeviceClass, then restarted the apiserver with the "bug". Reading the class works, updating it fails, leaving the class unmodified.

pohly commented:
I chatted with @liggitt about this. He still doesn't like that we are [proposing to make an exception](https://github.com/kubernetes/kubernetes/pull/129690#issuecomment-2598404158) for something that normally wouldn't qualify for backporting and said that he "won't approve, but won't block it".

@johnbelamaric: can you perhaps confirm that we want this backported by adding the LGTM? These are mechanical cherry-picks from master, so no new technical changes.

@thockin: can you approve (needed because it's touching some API files)?

/triage accepted
/priority important-soon

thockin commented:
The only things about this that make it viable to me are that it is beta and actually reported by implementors.  I am not sure what the guiding principle would be to say yes to this and no to something else, though.  It's not a "bug" per se, but it kind of is, right?

liggitt commented:
Yeah, thanks for the due diligence on containing the blast radius if there's an accidental bug in this code, and for verifying the behavior in the worst case scenario we can imagine where this new code completely panics. Both of those benefit master / 1.33+ as well, regardless of what we do here.

This is at the point now where we don't *think* this is risky to backport.

My reluctance comes from years of past experience where we (including I) would approve what we *thought* were risk-free backports, which later turned out to regress patch releases. From Kubernetes 1.19-1.27, [~35% of our patch releases](https://docs.google.com/spreadsheets/d/1LbGKBC4D2sLkcmzY9qDx9u-1D9TKC_ZrM8iA1eHW4Hs/edit?gid=751747416#gid=751747416) were regressed due to a backport and worse than the .0 of their minor in some way. That was a *terrible* track record, and taught users that picking up Kubernetes patch releases was somewhat of a gamble, stability-wise.

In 1.28+, when we got [*significantly* more strict about what we allow in backports](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-release/cherry-picks.md#what-kind-of-prs-are-good-for-cherry-picks), our track record has improved to ~8% of patch releases regressing (and is trending even better in newer minors). We need patch releases to be as close to ~0 risk as we can make them, so we can release critical bugfixes and security fixes and count on fast adoption.

Do I see a specific problem with this backport? Not really, though it is ~big and difficult to eyeball. Do I think that in aggregate, making exceptions and backporting things outside our narrow guidelines makes Kubernetes patch releases worse and less reliable? Yes, and I think the data supports that.

johnbelamaric commented:
I think the limited blast radius and the two independent implementors being limited by this justifies the backport.

/lgtm

thockin commented:
If I am left as the tie-breaker, I'll try to fight my risk-aversion instinct and bias toward action.

/approve
/hold  last chance to disagree?

pohly commented:
>  It's not a "bug" per se, but it kind of is, right?

It is a bug, just not a regression. Fixes for regressions justify a backport (can't have 1.(X+1) be worse than 1.X), fixed for bugs don't (new feature in 1.X simply not usable yet, users have to wait for 1.(X+1)).

Let me get release team approval, then I'll lift the hold.

pohly commented:
/hold cancel
